File: app/src/main/java/com/example/firsttestknocker/NotifAdapter.java
package com.example.firsttestknocker;


import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.Icon;
import android.net.Uri;
import android.support.constraint.ConstraintLayout;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.load.engine.Resource;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;

public class NotifAdapter extends BaseAdapter {
    private Context context;
    private View view;
    private ArrayList<StatusBarParcelable> notifications;
    private WindowManager windowManager;
    private final String TAG = NotificationListener.class.getSimpleName();
    private  ContactsRoomDatabase notification_listener_ContactsDatabase ;
    private  DbWorkerThread notification_listener_mDbWorkerThread ;
    private final String FACEBOOK_PACKAGE = "com.facebook.katana";
    private final String MESSENGER_PACKAGE = "com.facebook.orca";
    private final String WATHSAPP_SERVICE = "com.whatsapp";
    private final String GMAIL_PACKAGE = "com.google.android.gm";
    private final String  MESSAGE_PACKAGE = "com.google.android.apps.messaging";

    public NotifAdapter(Context context, ArrayList<StatusBarParcelable> listNotification, WindowManager windowManager,View view){
        this.context=context;
        this.notifications= listNotification;
        this.windowManager=windowManager;
        this.view= view;
        notification_listener_mDbWorkerThread = new DbWorkerThread("dbWorkerThread");
        notification_listener_mDbWorkerThread.start();
        //on get la base de données
        //notification_listener_ContactsDatabase = ContactsRoomDatabase.getDatabase(context);
    }

    public int getCount() {
        return notifications.size();
    }

    @Override
    public StatusBarParcelable getItem(int position) {
        return notifications.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, final ViewGroup parent) {
        if(convertView == null){
            convertView = LayoutInflater.from(context).inflate(R.layout.activity_notification_adapter,parent,false);
        }
        final StatusBarParcelable sbp = getItem(position);
        TextView app= (TextView) convertView.findViewById(R.id.notification_plateformeTv);
        TextView contenue= (TextView) convertView.findViewById(R.id.notification_contenu);
        ImageView appImg= (ImageView) convertView.findViewById(R.id.notification_plateformeImg);
        ImageView expImg= (ImageView) convertView.findViewById(R.id.notification_expediteurImg);
        final Button buttonResponse = (Button) convertView.findViewById(R.id.notification_adapter_button_response);
        final Button buttonSend = (Button) convertView.findViewById(R.id.notification_adapter_button_send);
        final EditText editText = (EditText) convertView.findViewById(R.id.notification_adapter_editText);

        app.setText(convertPackageToString(sbp.getAppNotifier()));
        contenue.setText(sbp.getStatusBarNotificationInfo().get("android.title")+":"+sbp.getStatusBarNotificationInfo().get("android.text"));
        //appImg.setImageResource(getApplicationNotifier(sbp));

        String pckg= sbp.getAppNotifier();
        if(sbp.getStatusBarNotificationInfo().get("android.icon")!=null) {
            int iconID = Integer.parseInt(sbp.getStatusBarNotificationInfo().get("android.icon").toString());
        }
        try {
            PackageManager pckManager = context.getPackageManager();
            Drawable icon = pckManager.getApplicationIcon(sbp.getAppNotifier());
            appImg.setImageDrawable(icon);
        }catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        if(!sbp.getStatusBarNotificationInfo().get("android.largeIcon").equals("")) {//image de l'expediteur provenant l'application source
            System.out.println("bitmap :"+sbp.getStatusBarNotificationInfo().get("android.largeIcon"));
            Bitmap bitmap = (Bitmap) sbp.getStatusBarNotificationInfo().get("android.largeIcon");
            expImg.setImageBitmap(bitmap);
        }
        View.OnClickListener listener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                System.out.println("click on constraint layout");
                String app = convertPackageToString(sbp.getAppNotifier());
                if (v.getId() == R.id.notification_adapter_button_response) {
                    System.out.println("click on button response");
                    buttonResponse.setVisibility(View.GONE);
                    buttonSend.setVisibility(View.VISIBLE);
                    editText.setVisibility(View.VISIBLE);
                    editText.requestFocus();
                    InputMethodManager inputMM= (InputMethodManager) context.getSystemService(Context.INPUT_METHOD_SERVICE);
                    inputMM.showSoftInput(editText,InputMethodManager.SHOW_IMPLICIT);
                } else if (v.getId() == R.id.notification_adapter_button_send){
                    System.out.println("click on button send");

                }else{

                    if (app.equals("Facebook")) {
                        gotToFacebookPage("");
                        closeNotification();
                    } else if (app.equals("Messenger")) {
                        gotToFacebookPage("");
                        closeNotification();
                    } else if (app.equals("WhatsApp")) {
                        onWhatsappClick();
                        closeNotification();
                    } else if (app.equals("gmail")) {
                        openGmail();
                        closeNotification();
                    } else if (app.equals("message")) {
                        openSms();
                        closeNotification();
                    }
                }


            }
        };

        contenue.setOnClickListener(listener);
        app.setOnClickListener(listener);
        app.setOnClickListener(listener);
        buttonResponse.setOnClickListener(listener);
        buttonSend.setOnClickListener(listener);
        return convertView;
    }
    private void closeNotification(){
        windowManager.removeView(view);
        SharedPreferences sharedPreferences = context.getSharedPreferences("Knocker_preferences", Context.MODE_PRIVATE);
        SharedPreferences.Editor edit= sharedPreferences.edit();
        edit.putBoolean("view",false);
        edit.commit();
    }
    private String  convertPackageToString(String packageName){
        if(packageName.equals(FACEBOOK_PACKAGE)){
            return "Facebook";
        }else if(packageName.equals(MESSENGER_PACKAGE)){
            return "Messenger";
        }else if(packageName.equals(WATHSAPP_SERVICE)){
            return "WhatsApp";
        }else if(packageName.equals(GMAIL_PACKAGE)){
            return "gmail";
        }else if(packageName.equals(MESSAGE_PACKAGE)){
            return "message";
        }
        return "";
    }
    private int getApplicationNotifier(StatusBarParcelable sbp ){

        if (sbp.getAppNotifier().equals(FACEBOOK_PACKAGE) || sbp.getAppNotifier().equals(MESSENGER_PACKAGE)) {
            return R.drawable.facebook;
        } else if (sbp.getAppNotifier().equals(GMAIL_PACKAGE)) {
            return R.drawable.gmail;
        } else if (sbp.getAppNotifier().equals(WATHSAPP_SERVICE)) {
            return R.drawable.download;
        }
        return R.drawable.sms;
    }


/////****** code dupliqué faire attention trouvé un moyen de ne plus en avoir *******//////
    private void gotToFacebookPage(String id) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://www.messenger.com/t/" + id));
            intent.setFlags(FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
        } catch (ActivityNotFoundException e) {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://www.messenger.com/t/" + id));
            context.startActivity(intent);
        }
    }


    private void onWhatsappClick(){
        //String url = "https://api.whatsapp.com/";
        // try{
            //PackageManager pm = context.getPackageManager();
            //pm.getPackageInfo("com.whatsapp", PackageManager.GET_ACTIVITIES);
            Intent i = context.getPackageManager().getLaunchIntentForPackage("com.whatsapp");
            i.setFlags(FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(i);
        /*} catch (PackageManager.NameNotFoundException e) {
            Toast.makeText(context, "Whatsapp app not installed in your phone", Toast.LENGTH_SHORT).show();
            e.printStackTrace();
        }*/

    }
    private void openGmail(){
        Intent i = context.getPackageManager().getLaunchIntentForPackage("com.google.android.gm");
        i.setFlags(FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(i);
    }
    private void openSms(){
        Intent i = context.getPackageManager().getLaunchIntentForPackage("com.google.android.apps.messaging");
        i.setFlags(FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(i);
    }

}

File: app/src/main/java/com/example/firsttestknocker/NotifAdapter.kt
package com.example.firsttestknocker


import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import java.util.ArrayList

import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.support.v4.content.ContextCompat.checkSelfPermission
import android.telephony.SmsManager

class NotifAdapter(private val context: Context, private val notifications: ArrayList<StatusBarParcelable>, private val windowManager: WindowManager, private val view: View) : BaseAdapter() {
    private val TAG = NotificationListener::class.java.simpleName
    private val notification_listener_ContactsDatabase: ContactsRoomDatabase? = null
    private val notification_listener_mDbWorkerThread: DbWorkerThread? = null
    private val FACEBOOK_PACKAGE = "com.facebook.katana"
    private val MESSENGER_PACKAGE = "com.facebook.orca"
    private val WATHSAPP_SERVICE = "com.whatsapp"
    private val GMAIL_PACKAGE = "com.google.android.gm"
    private val MESSAGE_PACKAGE = "com.google.android.apps.messaging"
    override fun getCount(): Int {
        return notifications.size
    }

    override fun getItem(position: Int): StatusBarParcelable {
        return notifications[position]
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        if (convertView == null) {
            convertView = LayoutInflater.from(context).inflate(R.layout.activity_notification_adapter, parent, false)
        }
        val sbp = getItem(position)
        val app = convertView!!.findViewById<View>(R.id.notification_plateformeTv) as TextView
        val contenue = convertView.findViewById<View>(R.id.notification_contenu) as TextView
        val appImg = convertView.findViewById<View>(R.id.notification_plateformeImg) as ImageView
        val expImg = convertView.findViewById<View>(R.id.notification_expediteurImg) as ImageView
        val buttonResponse = convertView.findViewById<View>(R.id.notification_adapter_button_response) as Button
        val buttonSend = convertView.findViewById<View>(R.id.notification_adapter_button_send) as Button
        val editText = convertView.findViewById<View>(R.id.notification_adapter_editText) as EditText

        app.text = convertPackageToString(sbp.appNotifier)
        contenue.text = sbp.statusBarNotificationInfo["android.title"].toString() + ":" + sbp.statusBarNotificationInfo["android.text"]
        //appImg.setImageResource(getApplicationNotifier(sbp));

        val pckg = sbp.appNotifier
        if (sbp.statusBarNotificationInfo["android.icon"] != null) {
            val iconID = Integer.parseInt(sbp.statusBarNotificationInfo["android.icon"]!!.toString())
        }
        try {
            val pckManager = context.packageManager
            val icon = pckManager.getApplicationIcon(sbp.appNotifier)
            appImg.setImageDrawable(icon)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }

        if (sbp.statusBarNotificationInfo["android.largeIcon"] != "") {//image de l'expediteur provenant l'application source
            println("bitmap :" + sbp.statusBarNotificationInfo["android.largeIcon"]!!)
            val bitmap = sbp.statusBarNotificationInfo["android.largeIcon"] as Bitmap?
            expImg.setImageBitmap(bitmap)
        }
        if(canResponse(sbp.appNotifier)){
            buttonResponse.visibility= View.VISIBLE
        }
        val listener = View.OnClickListener { v ->
            println("click on constraint layout")
            val app = convertPackageToString(sbp.appNotifier)
            if (v.id == R.id.notification_adapter_button_response) {
                println("click on button response")
                buttonResponse.visibility = View.GONE
                buttonSend.visibility = View.VISIBLE
                editText.visibility = View.VISIBLE
                editText.requestFocus()
                val inputMM = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                inputMM.showSoftInput(editText, InputMethodManager.SHOW_IMPLICIT)
            } else if (v.id == R.id.notification_adapter_button_send) {
                println("click on button send")
                val message = editText.text.toString()
                // String number= ContactInfo.
                val number= ContactInfo.getInfoWithName(sbp.statusBarNotificationInfo["android.title"].toString(),app)
                if(sbp.appNotifier.equals(MESSAGE_PACKAGE)){
                    val smsManager= SmsManager.getDefault()
                    smsManager.sendTextMessage(number,null, message,null,null)
                }
                println("message :"+message + "sendto : "+ number)
                closeNotification()

            } else {

                if (app == "Facebook") {
                    ContactGesture.openMessenger("",context)//TODO modifier si modification pour accès au post fb
                    closeNotification()
                } else if (app == "Messenger") {
                    ContactGesture.openMessenger("",context)
                    closeNotification()
                } else if (app == "WhatsApp") {
                    onWhatsappClick()
                    closeNotification()
                } else if (app == "gmail") {
                    ContactGesture.openGmail(context)
                    closeNotification()
                } else if (app == "message") {
                    openSms()
                    closeNotification()
                }
            }
        }

        contenue.setOnClickListener(listener)
        app.setOnClickListener(listener)
        app.setOnClickListener(listener)
        buttonResponse.setOnClickListener(listener)
        buttonSend.setOnClickListener(listener)
        return convertView
    }

    private fun closeNotification() {
        windowManager.removeView(view)
        val sharedPreferences = context.getSharedPreferences("Knocker_preferences", Context.MODE_PRIVATE)
        val edit = sharedPreferences.edit()
        edit.putBoolean("view", false)
        edit.commit()
    }

    private fun convertPackageToString(packageName: String): String {
        if (packageName == FACEBOOK_PACKAGE) {
            return "Facebook"
        } else if (packageName == MESSENGER_PACKAGE) {
            return "Messenger"
        } else if (packageName == WATHSAPP_SERVICE) {
            return "WhatsApp"
        } else if (packageName == GMAIL_PACKAGE) {
            return "gmail"
        } else if (packageName == MESSAGE_PACKAGE) {
            return "message"
        }
        return ""
    }
    private fun canResponse(packageName: String):Boolean{
        if((checkSelfPermission(context,Manifest.permission.SEND_SMS)==PackageManager.PERMISSION_GRANTED)&&(packageName== MESSAGE_PACKAGE || packageName==WATHSAPP_SERVICE)){
            return true
        }
        return false
    }
    private fun getApplicationNotifier(sbp: StatusBarParcelable): Int {

        if ((sbp.appNotifier == FACEBOOK_PACKAGE || sbp.appNotifier == MESSENGER_PACKAGE)) {
            return R.drawable.facebook
        } else if (sbp.appNotifier == GMAIL_PACKAGE) {
            return R.drawable.gmail
        } else if (sbp.appNotifier == WATHSAPP_SERVICE) {
            return R.drawable.download
        }
        return R.drawable.sms
    }


    /////****** code dupliqué faire attention trouvé un moyen de ne plus en avoir *******//////



    private fun onWhatsappClick() {
        //String url = "https://api.whatsapp.com/";
        // try{
        //PackageManager pm = context.getPackageManager();
        //pm.getPackageInfo("com.whatsapp", PackageManager.GET_ACTIVITIES);
        val i = context.packageManager.getLaunchIntentForPackage("com.whatsapp")
        i!!.flags = FLAG_ACTIVITY_NEW_TASK
        context.startActivity(i)
        /*} catch (PackageManager.NameNotFoundException e) {
            Toast.makeText(context, "Whatsapp app not installed in your phone", Toast.LENGTH_SHORT).show();
            e.printStackTrace();
        }*/

    }



    private fun openSms() {
        val i = context.packageManager.getLaunchIntentForPackage("com.google.android.apps.messaging")
        i!!.flags = FLAG_ACTIVITY_NEW_TASK
        context.startActivity(i)
    }

}

